### PR TITLE
feat(traffic_light_arbiter): basic initial implementation

### PR DESCRIPTION
**Pull request by knzo25**
## Description

This PR implements the arbiter mentioned in 
https://github.com/autowarefoundation/autoware.universe/issues/2567

Merging is done either via confidence or via V2X priority.

## Related links

Internal discussion 1: [TIER IV INTERNAL LINK](https://star4.slack.com/archives/C03S84LDJGG/p1683556646447669)
Internal discussion 2: [TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/CT/pages/2783118221/SP+2023-05-22+TLR)
Tracker: [TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-735)

## Tests performed
Ran a rosbag and observed the output of this node using as input the signals coming from `traffic_light_selector`

Rosbag: [TIER IV INTERNAL LINK](https://drive.google.com/drive/folders/1yXCFdeuCvtEMMd8KNSDKOHL8UPJNaoXi)

## Notes for reviewers
Without particular v2x signals, one way to check this node is:
 - Download some of the rosbags suggested in `Internal discussion 1`
 - Replay a rosbag mapping the camera to the old name: `ros2 bag play . -r 0.2 --clock --remap /tf_static:=/tf_static_old /sensing/camera/camera6/camera_info:=/sensing/camera/traffic_light/camera_info /sensing/camera/camera6/image_raw/compressed:=/sensing/camera/traffic_light/image_raw/compressed`
 - Launch the message converter: `ros2 launch traffic_light_selector traffic_light_selector.launch.xml`
 - Launch the arbiter: `ros2 launch traffic_light_arbiter traffic_light_arbiter.launch.xml`
 - Observe the output: `ros2 topic echo /perception/traffic_light_arbiter/traffic_signals`

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/traffic_light_arbiter/package.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  <version>0.1.0</version>
+  <description>The traffic_light_arbiter package</description>
+  <maintainer email="kenzo.lobos@tier4.jp">Kenzo Lobos-Tsunekawa</maintainer>
+  <license>Apache License 2.0</license>
+
+  <buildtool_depend>ament_cmake_auto</buildtool_depend>
+  <buildtool_depend>autoware_cmake</buildtool_depend>
+
+  <depend>autoware_auto_mapping_msgs</depend>
+  <depend>autoware_auto_perception_msgs</depend>
```
Comment (ktro2828) [inline perception/traffic_light_arbiter/package.xml:14]: [Q] Does this pacakge depend on `autoware_auto_perception_msgs`??

Comment (knzo25) [reply perception/traffic_light_arbiter/package.xml:14]: No, it slipped in when I copied a similars package's xml file. Thanks for the catch !

**Review by ktro2828 — COMMENTED** [review]
@knzo25 The codes look good to me :+1: 
However, the name of launch `traffic_light_selector.launch.xml` makes me confused.
I suppose that `traffic_light_arbiter.launch.xml` is natural. Moreover, there is another package named `traffic_light_selector`. Are there any reasons you named as this?

**Comment by knzo25** [comment]
Sorry, I had fixed the name of the launcher in [fd4fbee](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/fd4fbeeddf70b28465713fe6b3f4e44bbe059c2a) but forgot to push it :bow: 

Regarding the selector, it had been initially implemented by Takagi-san (https://github.com/autowarefoundation/autoware.universe/issues/2567#issuecomment-1469301672) as part of the old/new message transition.
That selector logic was only partial, hence the need of this PR. We still need the converter though, as the perception nodes have yet to transition to the new messages

**Review by miursh — APPROVED** [review]
LGTM
just with some minor comments

File: perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  if (
+    (rclcpp::Time(msg->stamp) - rclcpp::Time(latest_perception_msg_.stamp)).seconds() >
+    perception_time_tolerance_) {
+    latest_v2x_msg_.signals.clear();
+  }
+
+  arbiterAndPublish(msg->stamp);
+}
+
+void TrafficLightArbiter::arbiterAndPublish(const builtin_interfaces::msg::Time & stamp)
```
Comment (miursh) [inline perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp:108]: [nits]
Maybe verb (arbitrate) is preferred for function name

**Comment by miursh** [comment]
@Mingyu1991 @yukkysaito @TakaHoribe @isamu-takagi 
This PR is step4 (merge node of V2X and perception-output) of this [issue](https://github.com/autowarefoundation/autoware.universe/issues/2567#issuecomment-1465386646).
If you have any comment or concern, please let me know

**Comment by yukkysaito** [comment]
Can we change the name to `external` instead of limiting it to `V2X`? 
There is a possibility to use this IF as an external input from FOA or even Web tools.

File: perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp (hunk 2 above)
Comment (knzo25) [reply perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp]: Addressed in [1c2af21](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/1c2af21cbf6ebe39887f514fa45a19ef0e5b3a07)

**Comment by knzo25** [comment]
> Can we change the name to `external` instead of limiting it to `V2X`? There is a possibility to use this IF as an external input from FOA or even Web tools.

Addressed in [1c2af21](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/1c2af21cbf6ebe39887f514fa45a19ef0e5b3a07)

**Comment by isamu-takagi** [comment]
There are three stages to merging traffic light. Does this node only do the third one?
1. Merge same traffic light for different cameras.
2. Merge traffic lights to traffic signal (traffic light group).
3. Merge perception traffic signals and external traffic signals.

**Comment by knzo25** [comment]
> Merge same traffic light for different cameras.
Merge traffic lights to traffic signal (traffic light group).

Are being done in @Mingyu1991 's PR

**Comment by isamu-takagi** [comment]
>Are being done in @Mingyu1991 's PR
https://github.com/autowarefoundation/autoware.universe/pull/3442

@knzo25 Thank you, I got it.

File: perception/traffic_light_arbiter/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+It's currently a provisional implementation.
+
+### Inputs / Outputs
+
+#### Input
+
+| Name                             | Type                                         | Description                                              |
+| -------------------------------- | -------------------------------------------- | -------------------------------------------------------- |
+| ~/sub/vector_map                 | autoware_auto_mapping_msgs/msg/HADMapBin     | The vector map to get valid traffic signal ids.          |
+| ~/sub/perception_traffic_signals | autoware_perception_msgs::msg::TrafficSignal | The traffic signals from the image recognition pipeline. |
```
Comment (Mingyu1991) [inline perception/traffic_light_arbiter/README.md:19]:
```suggestion
| ~/sub/perception_traffic_signals | autoware_perception_msgs/msg/TrafficSignal | The traffic signals from the image recognition pipeline. |
```

File: perception/traffic_light_arbiter/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+| ~/sub/v2x_traffic_signals        | autoware_perception_msgs::msg::TrafficSignal | The traffic signals from the V2X interface.              |
```
Comment (Mingyu1991) [inline perception/traffic_light_arbiter/README.md:20]:
```suggestion
| ~/sub/v2x_traffic_signals        | autoware_perception_msgs/msg/TrafficSignal | The traffic signals from the V2X interface.              |
```

**Comment by Mingyu1991** [comment]
I think it would be better to follow the common practice that source files and include files are separated into different directories to avoid possible troubles when other packages are including this package in the future

File: perception/traffic_light_arbiter/README.md (hunk 3 above)
Comment (knzo25) [reply perception/traffic_light_arbiter/README.md:19]: Actually, `types` use `::` unlike include paths that use `/` .
Please refer to other packages (e.g., https://github.com/autowarefoundation/autoware.universe/tree/main/perception/lidar_centerpoint). 
Still, some of the other topics were wrong, so I fixed them in [271fcc0](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/271fcc0a9110feee40e8f53bc6e93cfdf0b06db1)

File: perception/traffic_light_arbiter/README.md (hunk 4 above)
Comment (knzo25) [reply perception/traffic_light_arbiter/README.md:20]: Actually, `types` use `::` unlike include paths that use `/` .
Please refer to other packages (e.g., https://github.com/autowarefoundation/autoware.universe/tree/main/perception/lidar_centerpoint). 
Still, some of the other topics were wrong, so I fixed them in [271fcc0](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/271fcc0a9110feee40e8f53bc6e93cfdf0b06db1)

**Comment by knzo25** [comment]
> I think it would be better to follow the common practice that source files and include files are separated into different directories to avoid possible troubles when other packages are including this package in the future

Done in [f0e3259](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/f0e32598a21867a18aabe234a9816eee0338a8ec)

**Review by Mingyu1991 — APPROVED** [review]
LGTM

**Comment by knzo25** [comment]
@miursh

File: perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  return signals;
+}
+
+}  // namespace lanelet
+
+TrafficLightArbiter::TrafficLightArbiter(const rclcpp::NodeOptions & options)
+: Node("traffic_light_selector", options)
+{
+  external_time_tolerance_ = this->declare_parameter<double>("external_time_tolerance", 5.0);
+  perception_time_tolerance_ = this->declare_parameter<double>("perception_time_tolerance", 1.0);
```
Comment (TakaHoribe) [inline perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp:51]: minor: these parameters seem to be different from the readme. Is it ok?

**Review by TakaHoribe — COMMENTED** [review]
@knzo25 Thank you. Overall, look good to me!

File: perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp (hunk 5 above)
Comment (knzo25) [reply perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp:51]: Addressed in [53b3540](https://github.com/autowarefoundation/autoware.universe/pull/3780/commits/53b35404ea1764c4f963bdd57e65ce7fc1db291f).
:bow:

Comment (TakaHoribe) [reply perception/traffic_light_arbiter/src/traffic_light_arbiter.cpp:51]: @knzo25 LGTM!

**Review by TakaHoribe — APPROVED** [review]

**Pull request merged**